controllers/serving/predictor.go

```diff
@@ -0,0 +1,165 @@
+		Strategy: v1.DeploymentStrategy{Type: v1.RollingUpdateDeploymentStrategyType},
+	}
+	deploy.Spec.Template.Labels = insertLabelsForPredictor(deploy.Spec.Template.Labels, inf, predictor)
+	// Set owner-reference to inference object.
+	err := controllerutil.SetControllerReference(inf, &deploy, ir.scheme)
+	if err != nil {
+		return nil, err
+	}
+	pvc, err := ir.buildModelVolume(modelVersion)
+	if err != nil {
+		return nil, err
+	}
+	// Search for existing volumes and check has model volume existed or not.
+	modelVolumeIndex := -1
+	for vi := range predictor.Template.Spec.Volumes {
+		v := &predictor.Template.Spec.Volumes[vi]
+		if v.PersistentVolumeClaim != nil && v.PersistentVolumeClaim.ClaimName == pvc.Name {
+			modelVolumeIndex = vi
+			break
+		}
+	}
+	if modelVolumeIndex < 0 {
+		predictor.Template.Spec.Volumes = append(predictor.Template.Spec.Volumes, corev1.Volume{
+			Name:         fmt.Sprintf("%s-volume", modelVersion.Name),
+			VolumeSource: corev1.VolumeSource{PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{ClaimName: pvc.Name}},
+		})
+		// Setup predictor template with specific framework requirements.
+		storageProvider := storage.GetStorageProvider(modelVersion.Spec.Storage)
+		path := storageProvider.GetModelPath(modelVersion.Spec.Storage)
+		setter := framework.Setters[inf.Spec.Framework]
+		if setter != nil {
+			// Setup predictor template by its specific framework type.
+			setter.SetSpec(&predictor.Template, modelVersion, path)
+		}
+
+		for ci := range predictor.Template.Spec.Containers {
+			predictor.Template.Spec.Containers[ci].VolumeMounts = append(predictor.Template.Spec.Containers[ci].VolumeMounts, corev1.VolumeMount{
+				Name:      fmt.Sprintf("%s-volume-mounts", modelVersion.Name),
+				MountPath: path,
+			})
+		}
+	}
+
+	// Create service to route traffic forward to predictors with specific model version and
+	// provide a load-balance frontend.
+	if err = ir.createServiceForPredictor(inf, predictor); err != nil {
+		return nil, err
+	}
+
+	if err = ir.client.Create(context.Background(), &deploy); err != nil {
+		return nil, err
+	}
+	return &deploy, nil
+}
+
+func (ir *InferenceReconciler) createServiceForPredictor(inf *v1alpha1.Inference, predictor *v1alpha1.PredictorSpec) error {
+	name := genPredictorName(inf, predictor)
+	predictorLabels := insertLabelsForPredictor(nil, inf, predictor)
+	svc := corev1.Service{}
+	if err := ir.client.Get(context.Background(), types.NamespacedName{Namespace: inf.Namespace, Name: name}, &svc); err == nil {
+		return nil
+	} else if !errors.IsNotFound(err) {
+		return err
+	}
+
+	// At this point, service not found and create a new one.
+
+	httpPort := getContainerPort(predictor, v1alpha1.DefaultPredictorContainerHttpPortName, v1alpha1.DefaultPredictorContainerHttpPort)
+	grpcPort := getContainerPort(predictor, v1alpha1.DefaultPredictorContainerGrpcPortName, v1alpha1.DefaultPredictorContainerGrpcPort)
+
+	svc = corev1.Service{
```

Comment:
what's the different between this service created vs the one created at inference_controller.go reconcile loop?

Reply:
there comes two-level of services in design:
1. `service` for inference-service: main entry of user traffic, host for users to access inference service.
2. `services` for each predictor: entry for each predictor deployment.

hence a user request will be routed as: 
```
requests ---> service(host: inference.default.svc)
                             |---> service1 (host: inference-predictor-model-1-1.default.svc)---> deployment 1.1
                             |---> service2 (host: inference-predictor-model-1-2.default.svc)---> deployment 1.2
```